File: Difficulty: Medium/Maximum Product Subarray/maximum-product-subarray.py
#User function Template for python3
class Solution:

	# Function to find maximum
	# product subarray
	def maxProduct(self,arr):
    	    
        if not arr:
            return 0
    
        max_ending_here = min_ending_here = max_so_far = arr[0]
    
        for num in arr[1:]:
            if num < 0:
                # Swap because multiplying by a negative flips max and min
                max_ending_here, min_ending_here = min_ending_here, max_ending_here
    
            max_ending_here = max(num, max_ending_here * num)
            min_ending_here = min(num, min_ending_here * num)
    
            max_so_far = max(max_so_far, max_ending_here)
    
        return max_so_far

		# code here


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    tc = int(input())
    while tc > 0:
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        ans = ob.maxProduct(arr)
        print(ans)
        tc -= 1

# } Driver Code Ends

File: Difficulty: Medium/Maximum Product Subarray/maximum-product-subarray.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>

using namespace std;


// } Driver Code Ends
// User function template for C++
class Solution {
  public:
  
    // Function to find maximum product subarray
    long long maxProduct(vector<int> &arr) {
        // code here
        vector<int> ps ;
        long long mx = -1e9 , pr=1;
        
        for(int i=0 ; i<arr.size() ; i++){
            
            if(arr[i]==0){
                pr=1;
                mx = max( mx , (long long)arr[i] );
            }
            else{
                pr*=arr[i];
                mx = max( mx , pr );
            }
            
        }
        
        pr=1;
        
        for(int i=arr.size()-1 ; i>=0 ; i--){
            
            if(arr[i]==0){
                pr=1;
                mx = max( mx , (long long)arr[i] );
            }
            else{
                pr*=arr[i];
                mx = max( mx , pr );
            }
            
        }
        
        return mx;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        // int n, i;
        // cin >> n;
        // vector<int> arr(n);
        // for (i = 0; i < n; i++) {
        //     cin >> arr[i];
        // }

        vector<int> arr;
        string input;

        // Read array
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        Solution ob;
        auto ans = ob.maxProduct(arr);
        cout << ans << "\n";

        cout << "~"
             << "\n";
    }
    return 0;
}
// } Driver Code Ends